proposals/p3848.md

```diff
@@ -0,0 +1,587 @@
+parameter list. Only the parameters that are named in the body will be read
+from, meaning the highest named parameter denotes the minimum number of
+arguments required by the function. The function body is free to omit
+lower-numbered parameters (ex: `fn { Print($10); }`).
+
+```
+// A lambda that takes two positional parameters being used as a comparator
+Sort(my_list, fn => $0.val < $1.val);
+```
+
+### Positional Parameter Restrictions
+
+**Proposal**: There are two restrictions applied to functions with positional
+parameters. The first restriction is that the definitions of function
+declarations must be attached to the declarations. The second restriction is
+that positional parameters can only be used in a context where there is exactly
+one enclosing function without an explicit parameter list. For example...
+
+```
+fn Foo1 {
+  fn Bar1 {}  // ❌ Invalid: Foo1 is already using positional parameters
+}
+
+fn Foo2 {
+  Print($0);
+  fn Bar2 {}  // ❌ Invalid: Foo2 is already using positional parameters
+}
+
+fn Foo3 {
+  fn Bar3 {
+    Print($0);  // ❌ Invalid: Foo3 is already using positional parameters
+  }
+}
+
+fn Foo4() {
+  fn Bar4 {
+    Print($0);  // ✅ Valid: Foo4 has explicit parameters
+  }
+}
+
+fn Foo5 {
+  fn Bar5() {}  // ✅ Valid: Bar5 has explicit parameters
+}
+
+fn Foo6() {
+  my_list.Sort(
+    fn => $0 < $1  // ✅ Valid: Foo6 has explicit parameters
+  );
+}
+```
+
+#### Alternative Considered: Additional Restriction
+
+In addition to the proposed restrictions, an additional restriction was
+considered. That being, visibility of functions with positional parameters could
+be restricted to only non-public interfaces. This alternative was considered by
+way of a leads question
+([#3860](https://github.com/carbon-language/carbon-lang/issues/3860)) and was
+decided against.
+
+## Function Captures
+
+**Proposal**: Function captures in Carbon mirror the non-init captures of C++. A
+function capture declaration consists of a capture mode (for `var` captures)
+followed by the name of a binding from the enclosing scope, and makes that
+identifier available in the inner function body. The lifetime of a capture is
+the lifetime of the function in which it exists. For example...
+
+```
+fn Foo() {
+  let handle: Handle = Handle.Get();
+  var thread: Thread = Thread.Make(fn [var handle] { handle.Process(); });
+  thread.Join();
+}
+```
+
+```
+fn Foo() {
+  let handle: Handle = Handle.Get();
+  fn MyThread[handle]() { handle.Process(); }
+  var thread: Thread = Thread.Make(MyThread);
+  thread.Join();
+}
+```
+
+### Capture Modes
+
+**Proposal**: `let` and `var` can appear as function captures. They behave as
+they would in regular bindings.
+
+To prevent ambiguities, captures can only exist on functions where the
+definition is attached to the declaration. This means they are supported on
+lambdas and they are supported on function declarations that are immediately
+defined, but they are not supported on forward-declared functions.
```

Comment:
Should this also mention only being supported inside the body of another function?

Or, alternatively, in an expression (lambda) or statement (function) context?

Reply:
Made this change to both captures and function fields sections :)